File: src/templates/cs/SeriesObject.cs
// ===============================================================================
// Authors: AFRL/RQQA
// Organization: Air Force Research Laboratory, Aerospace Systems Directorate, Power and Control Division
// 
// Copyright (c) 2017 Government of the United State of America, as represented by
// the Secretary of the Air Force.  No copyright is claimed in the United States under
// Title 17, U.S. Code.  All Other Rights Reserved.
// ===============================================================================

// This file was auto-created by LmcpGen. Modifications will be overwritten.

using System;
using System.IO;
using System.Text;
using System.Collections.Generic;
using Avtas.Lmcp;
-<import_series>-

-<series_namespace>- {

    /// <summary>
    /// -<object_comment>-
    /// </summary>
    public class -<object_type_name>- : -<parent_object_name>-
    {
        private const uint LMCP_TYPE = -<object_type_number>-;
private const string SERIES_NAME = -<series_name_setup>-;
private const ushort SERIES_VERSION = -<series_version>-;
/* Series name expressed as a long for quick comparisons */
private const long SERIES_ID = -<series_id>-;
private const string TYPE_NAME = "-<object_type_name>-";

        -<member_declaration>-

        /// <summary>
        /// Initializes a new instance of the <see cref="-<object_type_name>-"/> class.
        /// </summary>
        public -<object_type_name>-() {}

        /// <summary>
        /// Creates a new object that is a copy of the current instance.
        /// </summary>
        /// <returns>
        /// A new object that is a copy of this instance.
        /// </returns>
        public -<_override>- object Clone()
{
  return ( -<object_type_name>-) MemberwiseClone();
}

/// <summary>
/// Calculates the size of this message not including the header or the checksum.
/// </summary>
/// <returns>The size, in bytes, as a signed integer.</returns>
public -<_override>- int CalculateSize()
{
  -<calc_size>-
        }

/// <summary>
/// Unpacks data from a Stream.
/// </summary>
/// <param name="buf">The Stream that the data will be unpacked from.</param>
public -<_override>- void Unpack( Stream buf )
{
  LmcpBinaryReader br = new LmcpBinaryReader( buf );

  -<member_unpack>-
        }

/// <summary>
/// Packs the data into a Stream.
/// </summary>
/// <param name="buf">The Stream that the data will be packed into.</param>
public -<_override>- void Pack( Stream buf )
{
  LmcpBinaryWriter bw = new LmcpBinaryWriter( buf );

  -<member_pack>-
        }

/// <summary>
/// Returns a <see cref="T:System.String"/> that represents the current <see cref="T:System.Object"/>.
/// </summary>
/// <returns>
/// A <see cref="T:System.String"/> that represents the current <see cref="T:System.Object"/>.
/// </returns>
public override string ToString()
{
  StringBuilder buf = new StringBuilder();
  buf.Append( "Object Type -<object_type_name>- { \n" );

  -<member_print>-

  buf.Append( "}\n" );

  return buf.ToString();
}

/// <summary>
/// Determines whether the specified <see cref="T:System.Object"/> is equal to the current <see cref="T:System.Object"/>.
/// </summary>
/// <param name="obj">The <see cref="T:System.Object"/> to compare with the current <see cref="T:System.Object"/>.</param>
/// <returns>
/// true if the specified <see cref="T:System.Object"/> is equal to the current <see cref="T:System.Object"/>; otherwise, false.
/// </returns>
/// <exception cref="T:System.NullReferenceException">
/// The <paramref name="obj"/> parameter is null.
/// </exception>
public override bool Equals( object obj )
{
  if ( obj == null || GetType() != obj.GetType() )
    return false;

  -<equals_compare_object>-
        }

/// <summary>
/// Serves as a hash function for a particular type.
/// </summary>
/// <returns>
/// A hash code for the current <see cref="T:System.Object"/>.
/// </returns>
public override int GetHashCode()
{
  return base.GetHashCode();
}

/// <summary>
/// Converts the object to an XML format.
/// </summary>
/// <param name="ws">The whitespace to use for padding.</param>
/// <returns>The object in an XML format.</returns>
public -<_override>- string ToXml( string ws )
{
  StringBuilder buf = new StringBuilder();

  -<xml_write_object>-


            return buf.ToString();
}

/// <summary>
/// Gets the numeric identifier for the LMCP type.
/// </summary>
public -<_override>- uint LmcpType
{ get { return LMCP_TYPE; } }

/// <summary>
/// Gets the name of the series this object belongs to.
/// </summary>
public -<_override>- string SeriesName
{ get { return SERIES_NAME; } }

/// <summary>
/// Gets a numeric identifier for the series this object belongs to.
/// </summary>
public -<_override>- long SeriesNameAsLong
{ get { return SERIES_ID; } }

/// <summary>
/// Gets a version number for the object.
/// </summary>
public -<_override>- ushort Version
{ get { return SERIES_VERSION; } }

        -<member_gets_and_sets>-
    }
}

File: src/templates/cpp/SeriesObject.cpp
// ===============================================================================
// Authors: AFRL/RQQA
// Organization: Air Force Research Laboratory, Aerospace Systems Directorate, Power and Control Division
// 
// Copyright (c) 2017 Government of the United State of America, as represented by
// the Secretary of the Air Force.  No copyright is claimed in the United States under
// Title 17, U.S. Code.  All Other Rights Reserved.
// ===============================================================================

// This file was auto-created by LmcpGen. Modifications will be overwritten.

#include <cassert>
#include <limits>
#include <sstream>
#include <iomanip>
#include "-<series_dir>-/-<datatype_name>-.h"
-<include_sub_dependencies>-

-<open_namespace>-

   // Subscription string is namespace separated by '.' followed by type name
   const std::string -<datatype_name>-::Subscription = "-<longdatatype_name_dots>-";
   const std::string -<datatype_name>-::TypeName = "-<datatype_name>-";
   const std::string -<datatype_name>-::SeriesName = "-<series_name>-";
   const int64_t -<datatype_name>-::SeriesId = -<series_id>-LL;
   const uint16_t -<datatype_name>-::SeriesVersion = -<series_version>-;
   const uint32_t -<datatype_name>-::TypeId = -<datatype_id>-;
   
   bool is-<datatype_name>-(avtas::lmcp::Object* obj)
   {
      if(!obj) return false;
      if(obj->getSeriesNameAsLong() != -<series_id>-LL) return false;
      if(obj->getSeriesVersion() != -<series_version>-) return false;
      if(obj->getLmcpType() != -<datatype_id>-) return false;
      return true;
   }
   
   bool is-<datatype_name>-(std::shared_ptr<avtas::lmcp::Object>& obj)
   {
      if(!obj) return false;
      if(obj->getSeriesNameAsLong() != -<series_id>-LL) return false;
      if(obj->getSeriesVersion() != -<series_version>-) return false;
      if(obj->getLmcpType() != -<datatype_id>-) return false;
      return true;
   }
   
   std::vector< std::string > -<datatype_name>-Descendants()
   {
       std::vector< std::string > descendants;
       
       -<all_descendants>-
       return descendants;
   }
   
   -<datatype_name>-::-<datatype_name>-(void) : -<full_parent_datatype>-()
   {
      -<initialize_attributes>-
   }
     
   -<datatype_name>-::-<datatype_name>-(const -<datatype_name>- &that) : -<full_parent_datatype>-(that)
   {
        -<copy_initializer_list>-
   }
   
   -<datatype_name>- & -<datatype_name>-::operator=(const -<datatype_name>- &that)
   {
      if (this != &that)
      {
         -<full_parent_datatype>-::operator=(that);
         -<destroy_attributes>-
         -<copy_initializer_list>-
      }
      return *this;
   }

   -<datatype_name>-* -<datatype_name>-::clone() const
   {
        return new -<datatype_name>-(*this);
   }
   
   -<datatype_name>-::~-<datatype_name>-(void)
   {
      -<destroy_attributes>-
   }
  
   void -<datatype_name>-::pack(avtas::lmcp::ByteBuffer & buf) const
   {
      // Call parent's pack method
      -<full_parent_datatype>-::pack(buf);
      // Copy the class into the buffer
      -<pack_attributes>-
   }
   
   void -<datatype_name>-::unpack(avtas::lmcp::ByteBuffer & buf)
   {
      // Call parent's unpack method
      -<full_parent_datatype>-::unpack(buf);
      // Copy the buffer into the class
      -<unpack_attributes>-
   }

   uint32_t -<datatype_name>-::calculatePackedSize(void) const
   {
      uint32_t size = 0;
      size += -<full_parent_datatype>-::calculatePackedSize();
      -<calculate_packed_size>-
      return size;
   }

   std::string -<datatype_name>-::toString(int32_t depth) const
   {
      std::string indent(depth*3, ' ');
      std::ostringstream oss;
      oss << std::setprecision(std::numeric_limits<double>::max_digits10);
      oss << indent << "Object ( -<datatype_name>- ) {\n";
      indent = std::string((++depth)*3, ' ');
      -<tostring_attributes>-
      indent = std::string((--depth)*3, ' ');
      oss << indent << "}\n";
      return oss.str();
   }

   std::string -<datatype_name>-::toXML(int32_t depth)
   {
      std::string ws(depth*3, ' ');
      std::ostringstream str;
      str << std::setprecision(std::numeric_limits<double>::max_digits10);
      -<xml_write_object>-
      return str.str();
   }

   bool -<datatype_name>-::operator==(const -<datatype_name>- & that)
   {
      if( -<full_parent_datatype>-::operator!=(that) )
      {
          return false;
      }
      -<equals_attributes>-
   }

   bool -<datatype_name>-::operator!=(const -<datatype_name>- & that)
   {
      return( !(operator==(that)) );
   }

   -<gets_and_sets_implementation>-
-<close_namespace>-
